File: src/lib/api/backup.ts
import { invoke } from '@tauri-apps/api/core';
import { getCurrentLanguage } from '../i18n';

export type BackupMetadata = {
  version: string;
  created_at: string;
  notes_count: number;
  folders_count: number;
};

export type BackupSettings = {
  id: number;
  enabled: boolean;
  frequency: 'daily' | 'weekly' | 'monthly';
  backup_path: string | null;
  last_backup_at: string | null;
  max_backups: number;
  created_at: string;
  updated_at: string;
};

export type UpdateBackupSettingsInput = {
  enabled: boolean;
  frequency: 'daily' | 'weekly' | 'monthly';
  backup_path: string | null;
  max_backups: number;
};

/**
 * バックアップを作成
 * @param backupPath バックアップファイルを保存するディレクトリパス
 * @returns 作成されたバックアップファイルのパス
 */
export async function createBackup(backupPath: string): Promise<string> {
  const locale = getCurrentLanguage();
  return await invoke<string>('create_backup', { backupPath, locale });
}

/**
 * バックアップから復元
 * @param backupFile 復元するバックアップファイルのパス
 */
export async function restoreBackup(backupFile: string): Promise<void> {
  const locale = getCurrentLanguage();
  return await invoke<void>('restore_backup', { backupFile, locale });
}

/**
 * バックアップファイルのメタデータを読み取る
 * @param backupFile バックアップファイルのパス
 * @returns バックアップメタデータ
 */
export async function readBackupMetadata(backupFile: string): Promise<BackupMetadata> {
  const locale = getCurrentLanguage();
  return await invoke<BackupMetadata>('read_backup_metadata', { backupFile, locale });
}

/**
 * 自動バックアップ設定を取得
 */
export async function getBackupSettings(): Promise<BackupSettings> {
  const locale = getCurrentLanguage();
  return await invoke<BackupSettings>('get_backup_settings', { locale });
}

/**
 * 自動バックアップ設定を更新
 */
export async function updateBackupSettings(
  input: UpdateBackupSettingsInput
): Promise<BackupSettings> {
  const locale = getCurrentLanguage();
  return await invoke<BackupSettings>('update_backup_settings', { input, locale });
}

File: src-tauri/src/commands/backup.rs
use crate::db::models::{BackupSettings, UpdateBackupSettingsInput};
use crate::i18n::I18n;
use crate::services::backup::{BackupMetadata, BackupService};
use std::sync::Arc;
use tauri::State;

#[tauri::command]
pub async fn create_backup(
  backup_service: State<'_, Arc<BackupService>>,
  backup_path: String,
  locale: Option<String>,
) -> Result<String, String> {
  let _i18n = I18n::new(locale.as_deref().unwrap_or("en"));
  backup_service.create_backup(backup_path)
}

#[tauri::command]
pub async fn restore_backup(
  backup_service: State<'_, Arc<BackupService>>,
  backup_file: String,
  locale: Option<String>,
) -> Result<(), String> {
  let _i18n = I18n::new(locale.as_deref().unwrap_or("en"));
  backup_service.restore_backup(backup_file)
}

#[tauri::command]
pub async fn read_backup_metadata(
  backup_service: State<'_, Arc<BackupService>>,
  backup_file: String,
  locale: Option<String>,
) -> Result<BackupMetadata, String> {
  let _i18n = I18n::new(locale.as_deref().unwrap_or("en"));
  backup_service.read_backup_metadata(backup_file)
}

#[tauri::command]
pub async fn get_backup_settings(
  backup_service: State<'_, Arc<BackupService>>,
  locale: Option<String>,
) -> Result<BackupSettings, String> {
  let _i18n = I18n::new(locale.as_deref().unwrap_or("en"));
  backup_service.get_backup_settings()
}

#[tauri::command]
pub async fn update_backup_settings(
  backup_service: State<'_, Arc<BackupService>>,
  input: UpdateBackupSettingsInput,
  locale: Option<String>,
) -> Result<BackupSettings, String> {
  let _i18n = I18n::new(locale.as_deref().unwrap_or("en"));
  backup_service.update_backup_settings(input)
}
